### DISCUSSION TITLE
How to handle dynamic IDs in Contract Tests

### DISCUSSION THREAD

**Discussion by noelmccrory** [specmatic/specmatic#1077] [Ask For Help]
If I have an API that exposes the following endpoints:

```
POST /my-resource
PUT /my-resource/{id}
GET /my-resource/{id}
```

`POST /my-resource` will create a new record in the database with an autogenerated UUID. That UUID is then used in the PUT and GET operations. I am using examples in my open api spec to drive the contract tests. But I don't see how I can make the {id} part of the call dynamic.

I see two options:

1. Create a new resource. Then run a database script to update the ID. This will be complex and there are a number of tables with FKs.
2. Create a new resource. Then inform Specmatic to use the new ID when running the PUT and GET tests.

Is there any functionality that could help with Option 2? Thanks!

**Comment by burdeamit** [comment]
Thanks for reaching out @noelmccrory. 

Test data setup is indeed a real problem that we see often. In Java projects, [Specmatic has support for JDBC  stubbing](https://specmatic.in/updates/jdbc-stubbing-with-specmatic-contract-testing/). By stubbing out the DB we can fully control the details such as resource ID. If you are interested in trying out this capability please [reach out to us](https://specmatic.in/contact-us/).

Also, we have prioritised the support for [Open API Links](https://swagger.io/docs/specification/links/) in our backlog. This capability should allow you to use the dynamically generated ID in your HTTP POST call in subsequent API calls.

Please let us know if you have any more questions.

**Reply by noelmccrory** [reply]
Open API Links looks promising. I'll keep an eye out for that. Thanks.

For now, I proceeded with Option 1. I was able to use `ALTER TABLE my_table DISABLE TRIGGER ALL;` with Postgres to disable constraint checks, update IDs to match the IDs from examples in the API spec, then re-enable triggers.
